### PR TITLE
PD disagg with Mooncake Connector: GDN support (Qwen3.5) and MLA support (Deepseek-V4-Flash)

### PR DESCRIPTION
**Pull request by andakai**
## Purpose

Mooncake P/D disaggregation did not work correctly for hybrid KV-cache models such as Qwen3.5, where one model has both full-attention KV cache groups and GDN state groups represented as `MambaSpec`. Nixl now has supported GDN pd disagg here https://github.com/vllm-project/vllm/pull/41869, but mooncake connector lacks this implementation.

The previous Mooncake path treated request block IDs and registered cache regions too much like a single homogeneous KV cache. That breaks hybrid models because:

- full-attention KV pages and GDN/Mamba state slots have different block semantics;
- HMA returns per-KV-cache-group block IDs, but Mooncake flattened/reused them across regions;
- FA logical blocks may need expansion to physical kernel blocks, while GDN/Mamba state blocks should stay in state-slot space;
- blocks-first FA regions may need K/V splitting, while GDN state regions must stay whole;
- remote decode for MambaSpec-style state needs the N-1 prompt boundary so decode recomputes the last prompt token locally.

## Related PRs / Issues

- [#39482](https://github.com/vllm-project/vllm/pull/39482): initial Mooncake hybrid attention support for Qwen3.5. This PR builds on that direction and includes a coauthor trailer for the PR author.
- [#41869](https://github.com/vllm-project/vllm/pull/41869): NIXL P/D disaggregation support for Qwen3.5 GDN. This PR brings the corresponding direct MooncakeConnector path closer to that validated NIXL behavior.
- [#44456](https://github.com/vllm-project/vllm/pull/44456): broader KV-cache layout refactor around Mamba cache standardization; related but not the same request-level Mooncake P/D transfer fix.
- [#46334](https://github.com/vllm-project/vllm/pull/46334): Mooncake logical/physical block-ratio handling; related to physical block mapping, but this PR handles hybrid FA + GDN group semantics.
- [#46004](https://github.com/vllm-project/vllm/pull/46004): Mooncake shared KV metadata for DeepSeek V4 PP/PD; related Mooncake region metadata work, but for a different model/topology.

## What This PR Changes

This PR turns direct `MooncakeConnector` from a homogeneous KV-page transfer path into a cache-spec-aware transfer path for Qwen3.5-style hybrid FA + GDN models.

First, the worker no longer assumes that the first layer's attention backend represents the whole model. Qwen3.5 can expose GDN as the first backend, and treating that backend as ordinary attention reaches `GDNAttentionBackend.get_kv_cache_shape()`, which is not implemented. **The connector now resolves all current attention backends and initializes `TransferTopology` with `is_mamba=kv_cache_config.has_mamba_layers`.**

Second, memory registration is separated from transfer-region metadata. Mooncake still registers each shared backing allocation only once, but it preserves every logical transfer region, **including FA and GDN regions that share the same base allocation under HMA**. This removes the old assumption that all cache views have the same block dimension: FA regions can be physical/kernel-block based while GDN state regions stay in logical state-slot space.

Third, request transfer planning keeps KV-cache group identity all the way to descriptor construction. Full-attention regions consume full-attention group block IDs, GDN/MambaSpec regions consume their own group block IDs, and null placeholder blocks from aligned MambaSpec state groups are skipped instead of transferred as real state.

Finally, logical-to-physical block handling is applied only where it is valid. Attention group block IDs are expanded to kernel/physical block IDs when the logical block size differs from the selected kernel block size, while GDN/MambaSpec state groups are not expanded because their block IDs represent recurrent state slots rather than per-token KV pages. 

The scheduler path also mirrors the MambaSpec N-1 boundary used by NIXL because the P/D transfer path transfers KV/state, not the logits for the final prompt token. For a prompt with tokens `[t0, ..., tN-1]`, the prefiller therefore transfers recurrent state after processing through `tN-2`, and the decoder recomputes `tN-1` locally before generating the first output token. This lets the decode side obtain the logits for `tN-1` and advance local recurrent state consistently. If the prefiller instead transferred state after `tN-1`, the decoder would have the state needed to continue decoding but would still miss the logits needed to sample the first generated token unless logits were transferred as another cross-node artifact.

## Evaluation

GSM8K evaluation settings:

```text
configs: 1p1d, 2p2d, 4p4d, 4p1d, 1p4d
connector: mooncake or nixl
model: Qwen3.5-0.8B
tp_per_instance: 1
num_questions: 1319
num_shots: 5
max_tokens: 256
max_model_len: 4096
concurrency: 256
```

Mooncake runs used `--mooncake-protocol rdma`.

## Test Results

Qwen3.5-0.8B, GSM8K, 1319 questions, 5-shot, `max_tokens=256`, `concurrency=256`:

| Config | NIXL accuracy | NIXL questions/s | NIXL tokens/s | Mooncake accuracy | Mooncake questions/s | Mooncake tokens/s |
|---|---:|---:|---:|---:|---:|---:|
| 1p1d | 0.328279 | 45.085 | 5321.579 | 0.323730 | 53.965 | 6336.957 |
| 2p2d | 0.321456 | 64.392 | 7599.558 | 0.327521 | 80.357 | 9515.340 |
| 4p4d | 0.319181 | 64.366 | 7537.483 | 0.314632 | 89.781 | 10532.337 |
| 4p1d | 0.322214 | 40.346 | 4742.483 | 0.320697 | 49.535 | 5839.768 |
| 1p4d | 0.316149 | 61.295 | 7218.631 | 0.325246 | 76.481 | 8999.332 |

Standalone Qwen3.5-0.8B reference:

| Model | Config | Accuracy | Questions/s | Tokens/s |
|---|---|---:|---:|---:|
| Qwen3.5-0.8B | standalone | 0.330553 | 27.278 | 3218.657 |

Pure full-attention regression check, Qwen3-0.6B, Mooncake 1p1d:

| Code version | Accuracy | Invalid rate | Questions/s | Tokens/s |
|---|---:|---:|---:|---:|
| origin/main `52fbe1228` | 0.391964 | 0.002274 | 87.127 | 8764.035 |
| current branch | 0.391964 | 0.003033 | 83.734 | 8501.110 |

The Qwen3-0.6B result keeps accuracy identical to six decimals, so this PR did not show a pure full-attention Mooncake accuracy regression.

## Out of scope

- mamba2 test and other hybrid models tests.
- hetero-tp and pp.

## AI Assistance

AI assistance was used to help inspect, summarize, and draft this change. The submitting human reviewed the changed lines and ran the validation above.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/distributed/kv_transfer/kv_connector/v1/mooncake/mooncake_connector.py [hunk 1]
```diff
@@ -650,10 +732,15 @@ def get_num_new_matched_tokens(
             # Remote prefill: get all prompt blocks from remote.
             assert not self.is_kv_producer
             token_ids = request.prompt_token_ids or []
-            count = len(token_ids) - num_computed_tokens
+            count = self._mamba_prefill_token_count(len(token_ids)) - (
```
Comment (zhewenl) [inline vllm/distributed/kv_transfer/kv_connector/v1/mooncake/mooncake_connector.py]: `_mamba_prefill_token_count` can be bit confusing here, maybe naming it like `_get_remote_prefill_token_count`

**Review by zhewenl — APPROVED** [review]
overall LGTM, thanks for the effort @andakai! Could you also help test dsv4-flash/gpt-oss to ensure SWA/FA is function as expected?

File: vllm/distributed/kv_transfer/kv_connector/v1/mooncake/mooncake_connector.py (hunk 1 above)
Comment (andakai) [reply vllm/distributed/kv_transfer/kv_connector/v1/mooncake/mooncake_connector.py]: Thanks! I agree with this change. The previous name only contains "mamba" and can be confusing. I also change the name in NIXL connector for consistency.

**Comment by andakai** [comment]
Hi @zhewenl , I tested the gpt-oss and deepseek-v4-flash model with mooncake pd disagg. I found that deepseek-v4-flash has pretty low accuracy even in origin/main branch. I will try to figure out the reason first.
| Model | Run | Config | Accuracy |
|---|---|---:|---:|
| `gpt-oss-20b` | Standalone | `standalone` | 0.3161485974 |
| `gpt-oss-20b` | Current PR P/D | `4p4d` | 0.3115996967 |
| `gpt-oss-20b` | `origin/main` P/D | `4p4d` | 0.3206974981 |
| `deepseek-v4-flash` | Standalone | `standalone` | 0.9507202426 |
| `deepseek-v4-flash` | Current PR 1p1d | `1p1d` | 0.2463987870 |
| `deepseek-v4-flash` | `origin/main` 1p1d | `1p1d` | 0.0485216073 |

**Comment by andakai** [comment]
MLA differs from regular FA/GDN mainly because **its KV cache physical layout may not be compact and contiguous**. For regular FA, a KV block is usually a continuous K/V region; in HND layout Mooncake virtually splits it into K and V halves, so `kv_block_len = block_len // 2`. For GDN/Mamba-style state, it is not a K/V split attention cache, so the whole state region should be transferred: `kv_block_len = block_len`. But **DeepSeek V4 Flash uses an fp8 MLA packed layout: the physical distance between two blocks can be large because of kernel alignment, packed storage, metadata/scales, or shared allocation**, while the actual MLA KV payload that needs to be transferred is only `layer_spec.page_size_bytes`.

The fix is to separate two concepts in Mooncake: `block_len` is the **real address stride** between blocks, computed as `cache.stride(0) * cache.element_size()`, while `kv_block_len` is **the number of bytes to copy for one transfer region**. For MLA/SlidingWindowMLA, `kv_block_len = layer_spec.page_size_bytes`; for regular FA with HND K/V split, `kv_block_len = block_len // 2`; for Mamba/GDN or other non-split regions, `kv_block_len = block_len`. 

For example, DeepSeek V4 Flash may have `stride = 1,039,680 bytes` but an effective MLA payload of only `8,640 bytes`. **The correct transfer is `src = base + block_id * 1,039,680`, `len = 8,640`**. If `8,640` is used as the address stride, block 1 and later read from the wrong address; if `1,039,680` is used as the copy length, block 0 already copies too much and may overwrite unrelated packed regions on the decode side. The RDMA registration was also changed to register the full backing `untyped_storage()` and deduplicate it with `seen_storage_ptrs`, so accesses using the real stride remain inside registered memory while avoiding duplicate registrations for multiple views sharing the same storage.

Now the results are here:
| Model | Config | TP / Instances | Accuracy |
|---|---|---|---:|
| gpt-oss-20b | 4P4D RDMA | TP=1, 4P+4D | 0.2979529946929492 |
| gpt-oss-20b | 4P4D RDMA | TP=1, 4P+4D | 0.2964366944655042 |
| gpt-oss-20b | standalone pool | TP=1, 8 instances | 0.3062926459438969 |
| DeepSeek-V4-Flash | standalone pool | TP=4, 2 instances | 0.9507202426080363 |
| DeepSeek-V4-Flash | 1P1D RDMA | TP=4, 1P+1D | 0.9492039423805914 |
| Qwen3.5-0.8B | standalone | TP=1, 1 instance | 0.3305534495830174 |
| Qwen3.5-0.8B | 4P4D RDMA | TP=1, 4P+4D | 0.3214556482183472 |
| Qwen3-0.6B | standalone | TP=1, 1 instance | 0.3995451099317665 |
| Qwen3-0.6B | 4P4D RDMA | TP=1, 4P+4D | 0.3912054586808188 |

**Review by Dao007forever — APPROVED** [review]

**Comment by zhewenl** [comment]
@andakai thanks for the working on the fix for MLA, I think the CI failure seems related to this PR. Would you mind taking a look?

**Comment by andakai** [comment]
1. Updated `test_mooncake_connector.py` to set kv_block_len_per_layer in mocked workers. For FA blocks-first split cases, the value is block_len // 2.

2. Updated `test_mooncake_connector_hybrid_mamba.py` to expect the full backing storage size, matching the implementation that deduplicates by untyped_storage() and registers the whole storage.

3. I didn't reproduce the nixl ci failure. As I just changed a function's name in nixl connector. I don't think this is a real problem in this pr.

**Comment by zhewenl** [comment]
sounds good, seems like the NIXL test is failing constantly on main: https://buildkite.com/organizations/vllm/analytics/suites/ci-1/tests/ee273a3f-28cd-8925-93f2-b04d3064270c

**Review by ZJY0516 — APPROVED** [review]

**Comment by KonnyakuMatcha** [comment]
Hi, could you please share the complete commands and configurations used to test Qwen3.5 with Mooncake P/D disaggregation, including the prefiller, decoder, proxy commands? I’d like to reproduce the accuracy results reported in this PR. Thanks!

**Comment by andakai** [comment]
Hi @KonnyakuMatcha  of course. I just tested this script. You can have a try.
```
#!/usr/bin/env bash
set -euo pipefail

cd /root/vllm

MODEL=Qwen3.5-0.8B
DATA_DIR=/tmp/vllm_mooncake_4p4d_gsm8k/data
RUN_DIR=/tmp/vllm_mooncake_4p4d_gsm8k
CUDA12_LIB=$PWD/.venv/lib/python3.12/site-packages/nvidia/cuda_runtime/lib

PREFILL_API=(8010 8011 8012 8013)
BOOTSTRAP=(8998 8999 9000 9001)
DECODE_API=(8020 8021 8022 8023)
PIDS=()
PROXY_ARGS=(--host 127.0.0.1 --port 8000)

mkdir -p "$RUN_DIR/logs"
export LD_LIBRARY_PATH="$CUDA12_LIB${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
trap 'kill -INT "${PIDS[@]}" 2>/dev/null || true' EXIT

for i in {0..3}; do
    CUDA_VISIBLE_DEVICES=$i \
    VLLM_PORT=$((21000 + i * 100)) \
    VLLM_MOONCAKE_BOOTSTRAP_PORT=${BOOTSTRAP[$i]} \
    .venv/bin/vllm serve "$MODEL" \
        --port "${PREFILL_API[$i]}" \
        --language-model-only \
        --max-model-len 4096 \
        --disable-uvicorn-access-log \
        --kv-transfer-config \
        '{"kv_connector":"MooncakeConnector","kv_role":"kv_producer"}' \
        >"$RUN_DIR/logs/prefill$((i + 1)).log" 2>&1 &
    PIDS+=("$!")
    PROXY_ARGS+=(
        --prefill "http://127.0.0.1:${PREFILL_API[$i]}" "${BOOTSTRAP[$i]}"
    )
done

for i in {0..3}; do
    CUDA_VISIBLE_DEVICES=$((i + 4)) \
    VLLM_PORT=$((22000 + i * 100)) \
    .venv/bin/vllm serve "$MODEL" \
        --port "${DECODE_API[$i]}" \
        --language-model-only \
        --max-model-len 4096 \
        --disable-uvicorn-access-log \
        --kv-transfer-config \
        '{"kv_connector":"MooncakeConnector","kv_role":"kv_consumer"}' \
        >"$RUN_DIR/logs/decode$((i + 1)).log" 2>&1 &
    PIDS+=("$!")
    PROXY_ARGS+=(--decode "http://127.0.0.1:${DECODE_API[$i]}")
done

PYTHONUNBUFFERED=1 .venv/bin/python \
    examples/disaggregated/mooncake_connector/mooncake_connector_proxy.py \
    "${PROXY_ARGS[@]}" >"$RUN_DIR/logs/proxy.log" 2>&1 &
PIDS+=("$!")

for port in "${PREFILL_API[@]}" "${DECODE_API[@]}"; do
    until curl -fsS --max-time 2 "http://127.0.0.1:$port/health" \
        >/dev/null 2>&1; do
        sleep 2
    done
    echo "$port ready"
done

until rg -q "All prefiller instances are ready" "$RUN_DIR/logs/proxy.log"; do
    sleep 2
done
echo "Mooncake 4P4D ready"

TMPDIR="$DATA_DIR" .venv/bin/python tests/evals/gsm8k/gsm8k_eval.py \
    --host http://127.0.0.1 \
    --port 8000 \
    --num-questions 1319 \
    --num-shots 5 \
    --max-tokens 256 \
    --temperature 0 \
    --seed 42 \
    --save-results "$RUN_DIR/gsm8k_results.json" \
    2>&1 | tee "$RUN_DIR/logs/gsm8k_eval.log"
```

The gsm8k_results.json:
```
{
  "accuracy": 0.3191811978771797,
  "invalid_rate": 0.00530705079605762,
  "latency": 7.962140526000439,
  "questions_per_second": 165.65897018431087,
  "total_output_tokens": 154893,
  "tokens_per_second": 19453.688300802474,
  "num_questions": 1319,
  "num_shots": 5,
  "max_tokens": 256,
  "timestamp": 1784036973.5587723
}
```
